server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -3305,30 +3293,19 @@ public void restoreShard(
         );
         final Executor executor = threadPool.executor(ThreadPool.Names.SNAPSHOT);
         final BlobContainer container = shardContainer(indexId, snapshotShardId);
-        synchronized (ongoingRestores) {
-            if (store.isClosing()) {
-                restoreListener.onFailure(new AlreadyClosedException("store is closing"));
-                return;
-            }
-            if (lifecycle.started() == false) {
-                restoreListener.onFailure(new AlreadyClosedException("repository [" + metadata.name() + "] closed"));
-                return;
-            }
-            final boolean added = ongoingRestores.add(shardId);
-            assert added : "add restore for [" + shardId + "] that already has an existing restore";
```

Comment:
This assertion no longer exists with the changes. #101948 is about this assertion failure. Do we care? Maybe I am missing how the usage of refCount fix the double restore issue?

Reply:
IMO we don't care. We need to be sure to avoid _all_ concurrent recovery activity on the same shard path, not just restores, and protect against that today in the recovery subsystem (fundamentally using `org.elasticsearch.env.NodeEnvironment#shardLock`). This extra check for concurrent restores doesn't help, and because it's a little out of line with the stronger checks in recovery it seems it fails sometimes. I didn't put much time into working out exactly how it fails, it didn't seem worth it.